mlx/c/linalg.cpp

```diff
@@ -4,76 +4,189 @@
+    const mlx_stream s) {
+  try {
+    mlx_array_set_(
+        *res,
+        mlx::core::linalg::norm(
+            mlx_array_get_(a),
+            ord,
+            (axis ? std::make_optional(std::vector<int>(axis, axis + axis_num))
+                  : std::nullopt),
+            keepdims,
+            mlx_stream_get_(s)));
+  } catch (std::exception& e) {
+    mlx_error(e.what());
+    return 1;
+  }
+  return 0;
 }
-extern "C" mlx_array mlx_linalg_norm_ord(
-    mlx_array a,
-    mlx_string ord,
-    const int* axis,
-    size_t num_axis,
+extern "C" int mlx_linalg_norm_ord(
+    mlx_array* res,
+    const mlx_array a,
+    const char* ord,
+    const int* axis /* may be null */,
+    size_t axis_num,
     bool keepdims,
-    mlx_stream s) {
-  RETURN_MLX_C_ARRAY(mlx::core::linalg::norm(
-      a->ctx,
-      MLX_CPP_STRING(ord),
-      MLX_CPP_OPT_INTVEC(axis, num_axis),
-      keepdims,
-      s->ctx));
+    const mlx_stream s) {
+  try {
+    mlx_array_set_(
+        *res,
+        mlx::core::linalg::norm(
+            mlx_array_get_(a),
+            std::string(ord),
+            (axis ? std::make_optional(std::vector<int>(axis, axis + axis_num))
+                  : std::nullopt),
+            keepdims,
+            mlx_stream_get_(s)));
+  } catch (std::exception& e) {
+    mlx_error(e.what());
+    return 1;
+  }
+  return 0;
 }
-extern "C" mlx_array mlx_linalg_norm(
-    mlx_array a,
-    const int* axis,
-    size_t num_axis,
+extern "C" int mlx_linalg_norm(
+    mlx_array* res,
+    const mlx_array a,
+    const int* axis /* may be null */,
+    size_t axis_num,
     bool keepdims,
-    mlx_stream s) {
-  RETURN_MLX_C_ARRAY(mlx::core::linalg::norm(
-      a->ctx, MLX_CPP_OPT_INTVEC(axis, num_axis), keepdims, s->ctx));
+    const mlx_stream s) {
+  try {
+    mlx_array_set_(
+        *res,
+        mlx::core::linalg::norm(
+            mlx_array_get_(a),
+            (axis ? std::make_optional(std::vector<int>(axis, axis + axis_num))
+                  : std::nullopt),
+            keepdims,
+            mlx_stream_get_(s)));
+  } catch (std::exception& e) {
+    mlx_error(e.what());
+    return 1;
+  }
+  return 0;
 }
-extern "C" mlx_array mlx_linalg_pinv(mlx_array a, mlx_stream s) {
-  RETURN_MLX_C_ARRAY(mlx::core::linalg::pinv(a->ctx, s->ctx));
+extern "C" int
+mlx_linalg_pinv(mlx_array* res, const mlx_array a, const mlx_stream s) {
+  try {
+    mlx_array_set_(
+        *res, mlx::core::linalg::pinv(mlx_array_get_(a), mlx_stream_get_(s)));
+  } catch (std::exception& e) {
+    mlx_error(e.what());
+    return 1;
+  }
+  return 0;
 }
-extern "C" mlx_tuple_array_array mlx_linalg_qr(mlx_array a, mlx_stream s) {
-  RETURN_MLX_C_ARRAYPAIR(mlx::core::linalg::qr(a->ctx, s->ctx));
+extern "C" int mlx_linalg_qr(
+    mlx_array* res_0,
+    mlx_array* res_1,
+    const mlx_array a,
+    const mlx_stream s) {
+  try {
+    std::tie(mlx_array_get_(*res_0), mlx_array_get_(*res_1)) =
```

Comment:
I get:

```
MLX error: expected a non-empty mlx_array at mlx-swift2/Source/Cmlx/include/mlx/c/linalg.cpp:161
```

from calling like this:

```swift
public func qr(_ array: MLXArray, stream: StreamOrDevice = .default) -> (MLXArray, MLXArray) {
    var r0 = mlx_array_new()
    var r1 = mlx_array_new()

    mlx_linalg_qr(&r0, &r1, array.ctx, stream.ctx)

    return (MLXArray(r0), MLXArray(r1))
}
```

The `mlx_array_get()` doesn't seem like it is meant to be used on out params.

Reply:
fixed, thanks!